### PR TITLE
[Rebase & FF] 202405: Add more support for Windows Capsules

### PR DESCRIPTION
**Pull request by VivianNK**
## Description

Cherry-picked from the following commits:
621b31e009
ac8ba933fb
8256e15a5c
2a29c92335

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

N/A

## Integration Instructions

N/A

### CODE REVIEW

File: BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+##
+# UefiBuild Plugin that supports Window Capsule files based on the
+# Windows Firmware Update Platform spec.
+# Creates INF, Cat, and then signs it
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper.py]: Copyright (c) Microsoft Corporation.

For all files.

File: BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper_plug_in.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "scope": "global",
+  "name": "Windows Capsule Support Helper Functions",
+  "module": "WindowsCapsuleSupportHelper"
+}
```
Comment (os-d) [inline BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper_plug_in.json]: nit: newline at EOF

File: BaseTools/Scripts/basetools_scripts_bin_path_env.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "scope": "global",
+  "flags": ["set_path"]
+}
```
Comment (os-d) [inline BaseTools/Scripts/basetools_scripts_bin_path_env.json]: nit: add newline at EOF

File: MdeModulePkg/Include/Library/CapsuleLib.h [hunk 4]
```diff
@@ -50,6 +50,27 @@ ProcessCapsuleImage (
   IN EFI_CAPSULE_HEADER  *CapsuleHeader
   );
 
+// MSChange - Add a routine to allow firmware-specific handling for capsules
```
Comment (os-d) [inline MdeModulePkg/Include/Library/CapsuleLib.h]: Change to `MU_CHANGE` in all places

File: MdeModulePkg/Include/Library/CapsuleLib.h [hunk 5]
```diff
@@ -50,6 +50,27 @@ ProcessCapsuleImage (
   IN EFI_CAPSULE_HEADER  *CapsuleHeader
```
Comment (os-d) [inline MdeModulePkg/Include/Library/CapsuleLib.h:50]: Please remove the note in the commit message as it is no longer relevant.

File: MdeModulePkg/Include/Library/CapsulePersistLib.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
+/** @file -- CapsulePersistLib.h
+A public library interface for persisting Capsules across reset.
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _CAPSULE_PERSIST_LIB_H_
```
Comment (os-d) [inline MdeModulePkg/Include/Library/CapsulePersistLib.h]: nit: `CAPSULE_PERSIST_LIB_H_`

File: MdeModulePkg/MdeModulePkg.dec [hunk 7]
```diff
@@ -183,6 +184,12 @@
   ## @libraryclass  This library is used to report memory type information change
   ##                across a reboot.
   MemoryTypeInformationChangeLib|Include/Library/MemoryTypeInformationChangeLib.h
+  ## MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec:191]: add an empty line before this

File: MdeModulePkg/MdeModulePkg.dec [hunk 8]
```diff
@@ -163,6 +163,7 @@
   ##  @libraryclass  Provides services to access UEFI variable flash information.
   #
   VariableFlashInfoLib|Include/Library/VariableFlashInfoLib.h
+  ## MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec]: This can be dropped

File: MdeModulePkg/MdeModulePkg.dsc [hunk 9]
```diff
@@ -73,9 +73,11 @@
   TimerLib|MdePkg/Library/BaseTimerLibNullTemplate/BaseTimerLibNullTemplate.inf
   SerialPortLib|MdePkg/Library/BaseSerialPortLibNull/BaseSerialPortLibNull.inf
   CapsuleLib|MdeModulePkg/Library/DxeCapsuleLibNull/DxeCapsuleLibNull.inf
+  CapsulePersistLib|MdeModulePkg/Library/CapsulePersistLibNull/CapsulePersistLibNull.inf      ## MU_CHANGE
   PcdLib|MdePkg/Library/BasePcdLibNull/BasePcdLibNull.inf
   CustomizedDisplayLib|MdeModulePkg/Library/CustomizedDisplayLib/CustomizedDisplayLib.inf
   FrameBufferBltLib|MdeModulePkg/Library/FrameBufferBltLib/FrameBufferBltLib.inf
+
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dsc]: This can be dropped.

**Review by os-d — APPROVED** [review]

File: BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper.py (hunk 1 above)
Comment (VivianNK) [reply BaseTools/Plugin/WindowsCapsuleSupportHelper/WindowsCapsuleSupportHelper.py]: In a new commit or is it okay to just amend the old commit to follow the current copyright?

File: MdeModulePkg/Universal/Console/ConPlatformDxe/ConPlatform.c [hunk 10]
```diff
@@ -809,7 +809,7 @@ MatchUsbClass (
   DeviceSubClass = DevDesc.DeviceSubClass;
   DeviceProtocol = DevDesc.DeviceProtocol;
 
-  // MSCHANGE - BEGIN - Add support for IAD-type multi-function devices.
```
Comment (kuqin12) [inline MdeModulePkg/Universal/Console/ConPlatformDxe/ConPlatform.c]: Where was this one checked in? Should we merge this change into the same commit?

File: MdeModulePkg/Library/DxeCapsuleLibNull/DxeCapsuleLibNull.c [hunk 11]
```diff
@@ -2,6 +2,7 @@
   Null Dxe Capsule Library instance does nothing and returns unsupport status.
 
 Copyright (c) 2007 - 2019, Intel Corporation. All rights reserved.<BR>
+Copyright (C) Microsoft Corporation. All rights reserved.
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibNull/DxeCapsuleLibNull.c]:
```suggestion
Copyright (c) Microsoft Corporation.
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c [hunk 12]
```diff
@@ -554,7 +623,9 @@ ProcessTheseCapsules (
     }
   }
 
-  DEBUG ((DEBUG_INFO, "Updating the firmware ......\n"));
+  // MS_CHANGE - Printing to the screen is unacceptable in production FW.
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c]:
```suggestion
  // MU_CHANGE - Printing to the screen is unacceptable in production FW.
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c [hunk 13]
```diff
@@ -679,6 +753,12 @@ ProcessCapsules (
 {
   EFI_STATUS  Status;
 
+  // MS_CHANGE [BEGIN]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c]:
```suggestion
  // MU_CHANGE [BEGIN]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c [hunk 14]
```diff
@@ -690,14 +770,17 @@ ProcessCapsules (
...
+  Status = ProcessTheseCapsules (TRUE);
+  //
+  // Reboot System if required after all capsule processed
+  //
+  if (mNeedReset) {
+    DoResetSystem ();
   }
 
+  // }
+  // MS_CHANGE [END]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c]:
```suggestion
  // MU_CHANGE [END]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.inf [hunk 15]
```diff
@@ -4,6 +4,7 @@
 #  Capsule library instance for DXE_DRIVER module types.
 #
 #  Copyright (c) 2016 - 2021, Intel Corporation. All rights reserved.<BR>
+#  Copyright (C) Microsoft Corporation. All rights reserved.
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.inf]:
```suggestion
#  Copyright (c) Microsoft Corporation.
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 16]
```diff
@@ -1615,6 +1654,30 @@ ProcessCapsuleImage (
   return ProcessThisCapsuleImage (CapsuleHeader, NULL, NULL);
 }
 
+/** MS_CHANGE - START
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
/** MU_CHANGE - START
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 17]
```diff
@@ -1615,6 +1654,30 @@ ProcessCapsuleImage (
...
+  )
+{
+  if (SupportCapsuleImage (CapsuleHeader) != EFI_SUCCESS) {
+    return EFI_UNSUPPORTED;
+  }
+
+  return PersistCapsule (CapsuleHeader);
+}
+
+// MS_CHANGE - END
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
// MU_CHANGE - END
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 18]
```diff
@@ -1245,6 +1267,14 @@ ProcessFmpCapsuleImage (
     return EFI_SUCCESS;
   }
 
+  // MS_CHANGE [BEGIN]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
  // MU_CHANGE [BEGIN]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 19]
```diff
@@ -1245,6 +1267,14 @@ ProcessFmpCapsuleImage (
     return EFI_SUCCESS;
   }
 
+  // MS_CHANGE [BEGIN]
+  // ConnectAll to ensure
+  //    All the communication protocol required by driver in capsule installed
+  //    All FMP protocols are installed
+  //
+  EfiBootManagerConnectAll ();
+  // MS_CHANGE [END]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
  // MU_CHANGE [END]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 20]
```diff
@@ -1270,6 +1300,15 @@ ProcessFmpCapsuleImage (
     }
   }
 
+  // MS_CHANGE [BEGIN]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
  // MU_CHANGE [BEGIN]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 21]
```diff
@@ -1270,6 +1300,15 @@ ProcessFmpCapsuleImage (
...
   }
 
+  // MS_CHANGE [BEGIN]
+  // Connnect all again to connect drivers within capsule
+  //
+  if (FmpCapsuleHeader->EmbeddedDriverCount > 0) {
+    EfiBootManagerConnectAll ();
+  }
+
+  // MS_CHANGE [END]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]:
```suggestion
  // MU_CHANGE [END]
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c [hunk 22]
```diff
@@ -251,6 +253,16 @@ ValidateFmpCapsule (
 
   FmpCapsuleHeaderSize = sizeof (EFI_FIRMWARE_MANAGEMENT_CAPSULE_HEADER) + sizeof (UINT64)*ItemNum;
 
+  // MS_CHANGE [BEGIN]
```
Comment (kuqin12) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c]: Same for these 4 instances :)

File: MdeModulePkg/Universal/Console/ConPlatformDxe/ConPlatform.c (hunk 10 above)
Comment (kuqin12) [reply MdeModulePkg/Universal/Console/ConPlatformDxe/ConPlatform.c]: I added the suggested MS_CHANGE to MU_CHANGE to other places. But I think we should drop this one as it is a comment only change and not needed to be added as a separate commit.

**Review by kuqin12 — APPROVED** [review]

File: MdeModulePkg/Include/Library/CapsuleLib.h [hunk 23]
```diff
@@ -50,6 +50,27 @@ ProcessCapsuleImage (
...
+//           that need to persist across reset.
+
+/**
+  The firmware-specific implementation that stages the capsule image
+  for processing after reset if it recognized the format of this capsule
+  image.
+
+  Caution: This function may receive untrusted input.
+
+  @param  CapsuleHeader    Pointer to the UEFI capsule image to be processed.
```
Comment (makubacki) [inline MdeModulePkg/Include/Library/CapsuleLib.h]: Could you please add the direction to the parameters in this commit?

```suggestion
  @param[in]  CapsuleHeader    Pointer to the UEFI capsule image to be processed.
```

File: MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLibNull.c [hunk 24]
```diff
@@ -68,3 +68,25 @@ ProcessCapsules (
 {
   return EFI_UNSUPPORTED;
 }
+
+// MS_CHANGE [BEGIN] - This implementation is BootServices/Runtime specific.
```
Comment (makubacki) [inline MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLibNull.c]: Based on existing comments, I know you're going to look at this now, but this is another instance.

```suggestion
// MU_CHANGE [BEGIN] - This implementation is BootServices/Runtime specific.
```

**Review by makubacki — APPROVED** [review]

**Pull request merged**